deepchem/models/tensorgraph/tensor_graph.py

```diff
@@ -0,0 +1,339 @@
+        feed_dict = {}
+        for self.epoch in range(self.epoch, self.epoch + nb_epoch):
+          avg_loss, n_batches = 0., 0
+          for ind, (X_b, y_b, w_b, ids_b) in enumerate(
+              dataset.iterbatches(batch_size, pad_batches=True)):
+            if ind % log_every_N_batches == 0:
+              print("On batch %d" % ind)
+            feed_dict = self._construct_feed_dict(X_b, y_b, w_b, ids_b)
+            output_tensors = [x.out_tensor for x in self.outputs]
+            fetches = output_tensors + [train_op, self.loss.out_tensor]
+            fetched_values = sess.run(fetches, feed_dict=feed_dict)
+            loss = fetched_values[-1]
+            avg_loss += loss
+            n_batches += 1
+          if self.epoch % checkpoint_interval == checkpoint_interval - 1:
+            saver.save(sess, self.save_file, global_step=self.epoch)
+            self._log_tensorboard(sess, feed_dict)
+          avg_loss = float(avg_loss) / n_batches
+          print('Ending epoch %d: Average loss %g' % (self.epoch, avg_loss))
+        self.last_checkpoint = saver.last_checkpoints[-1]
+      ############################################################## TIMING
+      time2 = time.time()
+      print("TIMING: model fitting took %0.3f s" % (time2 - time1))
+      ############################################################## TIMING
+
+  def _log_tensorboard(self, sess, feed_dict):
+    """
+    TODO(LESWING) set epoch
+    Parameters
+    ----------
+    sess
+    feed_dict
+
+    Returns
+    -------
+
+    """
+    if not self.tensorboard:
+      return
+    summary = sess.run(self._get_tf("summary_op"), feed_dict=feed_dict)
+    print("Loggin")
+    writer = self._get_tf("FileWriter")
+    writer.reopen()
+    writer.add_summary(summary, global_step=self.epoch)
+    writer.close()
+
+  def fit_on_batch(self, X, y, w):
+    if not self.built:
+      self.build()
+    dataset = NumpyDataset(X, y)
+    return self.fit(dataset, nb_epoch=1)
+
+  def _construct_feed_dict(self, X_b, y_b, w_b, ids_b):
+    feed_dict = dict()
+    if len(self.labels) > 0 and y_b is not None:
+      feed_dict[self.labels[0].out_tensor] = y_b
+    if len(self.features) > 0 and X_b is not None:
+      feed_dict[self.features[0].out_tensor] = X_b
+    return feed_dict
+
+  def predict_on_batch(self, X):
+    """Generates output predictions for the input samples,
+      processing the samples in a batched way.
+
+    # Arguments
+        x: the input data, as a Numpy array.
+        batch_size: integer.
+        verbose: verbosity mode, 0 or 1.
+
+    # Returns
+        A Numpy array of predictions.
+    """
+    if len(self.features) != 1:
+      raise ValueError("Only allow one input set of features")
+    features = self.features[0]
+    if not self.built:
+      self.build()
+    with self.tensor_objects['Graph'].as_default():
+      saver = tf.train.Saver()
+      with tf.Session() as sess:
+        saver.restore(sess, self.last_checkpoint)
```

Comment:
Restoring on every batch is going to get pretty slow. I'd recommend also directly overriding the `predict()` method so that only one restore is needed for large `predict()` operations.